drivers/SmartThings/zigbee-thermostat/src/popp_danfoss/init.lua

```diff
@@ -0,0 +1,109 @@
+
+local BAT_MIN = 24.0
+local BAT_MAX = 32.0
+
+local POPP_THERMOSTAT_FINGERPRINTS = {
+  { mfr = "D5X84YU", model = "eT093WRO" },
+  { mfr = "Danfoss", model = "eTRV0100" }
+}
+
+local is_popp_thermostat = function(opts, driver, device)
+  for _, fingerprint in ipairs(POPP_THERMOSTAT_FINGERPRINTS) do
+    if device:get_manufacturer() == fingerprint.mfr and device:get_model() == fingerprint.model then
+      return true
+    end
+  end
+  return false
+end
+
+local supported_thermostat_mode_handler = function(driver, device, supported_modes)
+  device:emit_event(ThermostatMode.supportedThermostatModes({"heat"}))
+end
+
+local thermostat_heating_setpoint_handler = function(driver, device, value)
+  device:emit_event(capabilities.thermostatHeatingSetpoint.heatingSetpoint({value = value.value/100.0, unit = "C"}))
+end
+
+local function set_heating_setpoint(driver, device, command)
+  local value = command.args.setpoint
+
+  if value >= 35 then -- we got a command in fahrenheit
+    value = utils.f_to_c(value)
+  end
+
+  if value >= DEFAULT_MIN_SETPOINT and value <= DEFAULT_MAX_SETPOINT then
+    device:send(Thermostat.attributes.OccupiedHeatingSetpoint:write(device, value * 100))
+    device:send(Thermostat.attributes.OccupiedHeatingSetpoint:read(device))
+  end
+end
+
+local battery_perc_attr_handler = function(driver, device, value, zb_rx)
+  device:emit_event(capabilities.battery.battery(utils.clamp_value(value.value, 0, 100)))
+end
+
+local battery_voltage_handler = function(driver, device, battery_voltage)
+  if (battery_voltage.value == 0) then -- this means we're plugged in
+    device:emit_event(PowerSource.powerSource.mains())
+    device:emit_event(Battery.battery(100))
+  else
+    local perc_value = utils.round((battery_voltage.value - BAT_MIN)/(BAT_MAX - BAT_MIN) * 100)
+    device:emit_event(Battery.battery(utils.clamp_value(perc_value, 0, 100)))
+  end
+end
+
+local set_thermostat_mode = function(driver, device, command)
+    device:send(Thermostat.attributes.SystemMode:write(device, ThermostatSystemMode.HEAT))
+    device.thread:call_with_delay(2, function(d)
+      device:send(Thermostat.attributes.SystemMode:read(device))
+    end)
+end
+
+local do_configure = function(self, device)
+  device:configure()
+  device:send(PowerConfiguration.attributes.BatteryVoltage:configure_reporting(device, 30, 21600, 1))
+end
+
+local popp_danfoss_thermostat = {
+  NAME = "POPP Danfoss Thermostat Handler",
+  zigbee_handlers = {
+    attr = {
+      [Thermostat.ID] = {
+        [Thermostat.attributes.ControlSequenceOfOperation.ID] = supported_thermostat_mode_handler,
+        [Thermostat.attributes.OccupiedHeatingSetpoint.ID] = thermostat_heating_setpoint_handler
+      },
+      [PowerConfiguration.ID] = {
+        [PowerConfiguration.attributes.BatteryPercentageRemaining.ID] = battery_perc_attr_handler
+      },
+      [PowerConfiguration.ID] = {
+        [PowerConfiguration.attributes.BatteryVoltage.ID] = battery_voltage_handler
+      }
```

Comment:
Those might be handled by the defaults. You might only need to add some configuration to the `init` lifecycle

`init = battery_defaults.build_linear_voltage_init(<bat_min>, <bat_max>),

Reply:
> Those might be handled by the defaults. You might only need to add some configuration to the `init` lifecycle
> 
> `init = battery_defaults.build_linear_voltage_init(<bat_min>, <bat_max>),

Didn't work like that `init = battery_defaults.build_linear_voltage_init(<bat_min>, <bat_max>)'. Unit tests were failing when I coded like that.
I think because of the re-setting this function in main driver. There is battery_voltage_handler implemented here:
https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/d4d8fac96878a2418f9b6e3e48ac9bc0532030c4/drivers/SmartThings/zigbee-thermostat/src/init.lua#L95

Reply:
You may try something like: 
```
local function device_init(driver, device)
...
...
  battery_defaults.build_linear_voltage_init(.......)(driver, device)
end
```